assets/src/scripts/components/hydrograph/legend.js

```diff
@@ -0,0 +1,122 @@
+                          textYPosition=0,
+                          markerGroupOffset=40,
+                          markerTextOffset=10) {
+    let legend = svg
+        .append('g')
+        .attr('class', 'legend');
+
+    let svgBBox = svg.node().getBBox();
+
+    let previousMarkerGroup;
+
+    for (let legendMarker of legendMarkers) {
+        let xPosition;
+        let previousMarkerGroupBox;
+        let detachedMarker;
+        if (previousMarkerGroup == null) {
+            xPosition = startingXPosition;
+        }
+        else {
+            previousMarkerGroupBox = previousMarkerGroup.node().getBBox();
+            xPosition = previousMarkerGroupBox.x + previousMarkerGroupBox.width + markerGroupOffset;
+        }
+        let markerType = legendMarker.type;
+        let legendGroup = legend.append('g')
+            .attr('class', 'legend-marker');
+        if (legendMarker.groupId) {
+            legendGroup.attr('id', legendMarker.groupId);
+        }
+        let markerArgs = {
+            r: legendMarker.r ? legendMarker.r : null,
+            x: xPosition,
+            y: markerYPosition,
+            length: 20,
+            domId: legendMarker.domId,
+            domClass: legendMarker.domClass
+        };
+        // add the marker to the svg
+        detachedMarker = markerType(markerArgs);
+        legendGroup.node().appendChild(detachedMarker.node());
+        // add text for the legend marker
+        let detachedMarkerBBox = detachedMarker.node().getBBox();
+        legendGroup.append('text')
+            .attr('x', detachedMarkerBBox.x + detachedMarkerBBox.width + markerTextOffset)
+            .attr('y', textYPosition)
+            .text(legendMarker.text);
+
+        previousMarkerGroup = legendGroup;
+    }
+    // center the legend group in the svg
+    let legendBBox = legend.node().getBBox();
+    let legendXPosition = (svgBBox.width - legendBBox.width) / 2;
+    legend.attr('transform', `translate(${legendXPosition}, ${svgBBox.height-15})`);
+}
+
+
+const legendDisplaySelector = createSelector(
+    (state) => state.showSeries,
+    (state) => state.statisticalMetaData,
+    (showSeries, statisticalMetaData) => {
+        let shownSeries = [];
+        let displayMarkers = [];
+        let text;
+        let marker;
+        for (let key in showSeries) {
+            if (showSeries.hasOwnProperty(key)) {
+                if (showSeries[key]) {
+                    shownSeries.push(key);
+                }
+            }
+        }
+        for (let seriesName of shownSeries) {
+            if (seriesName === 'compare' || seriesName === 'current') {
+                text = 'Current Year';
+                let domId = `ts-${seriesName}`;
+                let svgGroup = `${seriesName}-line-marker`;
+                if (seriesName === 'compare') {
+                    text = 'Last Year';
+                }
+                marker = defineLineMarker(domId, 'line', text, svgGroup);
```

Comment:
I'm not sure, but I kind of feel like the actual DOM nodes should be created by the `drawSimpleLegend` function and the selectors just used for grabbing the state necessary to create the elements. But maybe that's not a distinction we should make? Selectors are pretty useful to use everywhere...

But to complete my thinking, what I mean is in `drawSimpleLegend` you could do something like:

```javascript
legendGroup.append(lineMarker('line', ....
// or maybe...
legendGroup.call(appendLineMarker, domId, 'line', text, svgGroup)
```

I'm curious your thoughts on that...

Reply:
So, I was thinking about couple things when I designed that:

* I wanted `drawSimpleLegend` to have absolutely no what legend markers it had to work with. I was thinking that this would make it more general for when the legend gets more complicated. Though, the logic to create the markers has to live somewhere, so perhaps that's not a pertinent concern.
* My understanding of the a selector it's used to compute things from the state. So, I thought that since the DOM nodes that would appear in the legend are determined by the `showSeries` attribute in the state, it seemed to make sense to derive them in them in the selector.

Reply:
Yes, it does work pretty well...

@mbucknell you have a thought? My gut is to keep all the dom creation in the functions you can trace directly off the main rendering function.

Reply:
In the spirit of MVC, it seems like the selectors are part of the Model, whereas DOM is clearly part of the View. I haven't looked at this part of the code in detail but IMHO we should avoid DOM manipulation in the selector. Of course, I think this code merely creates the nodes and they are placed later. So I'm leaning more towards having the DOM nodes created in the rendering function rather than the selector.

Reply:
Sure. Perhaps we can have the selector return an array of the things that need a legend. Then `drawSimpleLegend` function can generate the DOM nodes for those markers before appending them to svg legend group?

Reply:
That seems good to me.

Reply:
@danielnaab -- Reworked. A bit different than what I had described, but it think it's cleaner this way.